File: 242-valid-anagram/solution.cpp
#include <string>
#include <unordered_map>

class Solution
{
private:
  static std::unordered_map<char, size_t> counter(const std::string &s)
  {
    std::unordered_map<char, size_t> s_counter;
    for (const char c : s)
    {
      ++s_counter[c];
    }

    return s_counter;
  }

public:
  bool isAnagram(std::string s, std::string t)
  {
    if (s.size() != t.size())
    {
      return false;
    }

    const auto s_counter = Solution::counter(s);
    const auto t_counter = Solution::counter(t);

    for (const auto &it : t_counter)
    {
      if (const auto s_it = s_counter.find(it.first); s_it == s_counter.end() || s_it->second != it.second)
      {
        return false;
      }
    }

    return true;
  }
};

File: 242-valid-anagram/solution.py
from collections import Counter


class Solution:
  def isAnagram(self, s: str, t: str) -> bool:
    if len(s) != len(t):
      return False

    s_counter = Counter(s)
    t_counter = Counter(t)

    for c, count in t_counter.items():
      if c not in s_counter or s_counter[c] != count:
        return False

    return True
